File: ios/ClaudeCursor/ClaudeCursor/Views/TerminalWebView.swift
import SwiftUI
import WebKit

struct TerminalWebView: UIViewRepresentable {
    let baseURL: URL
    let sessionName: String
    var sharePath: String? = nil

    /// Shared reference so KeyboardToolbar can inject keys
    static var sharedWebView: WKWebView?

    func makeCoordinator() -> Coordinator {
        Coordinator()
    }

    func makeUIView(context: Context) -> WKWebView {
        let config = WKWebViewConfiguration()
        config.allowsInlineMediaPlayback = true

        let webView = WKWebView(frame: .zero, configuration: config)
        webView.isOpaque = false
        webView.backgroundColor = .black
        webView.scrollView.isScrollEnabled = true
        webView.scrollView.bounces = false
        webView.scrollView.contentInsetAdjustmentBehavior = .never
        webView.scrollView.delaysContentTouches = false
        webView.navigationDelegate = context.coordinator

        // Include "Safari" in UA to fix xterm.js rendering in WKWebView
        webView.customUserAgent = "Mozilla/5.0 (iPhone; CPU iPhone OS 17_0 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/17.0 Mobile/15E148 Safari/604.1 ClaudeCursorCompanion/1.0"

        context.coordinator.currentSession = sessionName
        Self.sharedWebView = webView

        loadTerminal(webView)
        return webView
    }

    func updateUIView(_ webView: WKWebView, context: Context) {
        if context.coordinator.currentSession != sessionName {
            context.coordinator.currentSession = sessionName
            loadTerminal(webView)
        }
    }

    private func loadTerminal(_ webView: WKWebView) {
        let urlString = baseURL.absoluteString.trimmingCharacters(in: CharacterSet(charactersIn: "/"))
        let path = sharePath ?? "/terminal/\(sessionName)/"
        guard let terminalURL = URL(string: "\(urlString)\(path)") else { return }
        webView.load(URLRequest(url: terminalURL))
    }

    class Coordinator: NSObject, WKNavigationDelegate {
        var currentSession: String?

        /// Modifier handler JS injected after page load
        private static let setupJS = """
        (function(){
            if(window._appSetup) return;
            window._appSetup=true;
            window._pendingModifier=null;

            /* Ctrl/Alt modifier handler — capture phase on document fires before xterm.js */
            document.addEventListener('keydown',function(e){
                if(!window._pendingModifier||e.key.length!==1) return;
                e.preventDefault();
                e.stopImmediatePropagation();
                var mod=window._pendingModifier;
                window._pendingModifier=null;
                if(window.term){
                    if(mod==='ctrl'){
                        var c=e.key.toUpperCase().charCodeAt(0)-64;
                        if(c>0&&c<27){var ch=String.fromCharCode(c);if(window._wsSend)window._wsSend(ch);else if(window.term)window.term.input(ch);}
                    } else if(mod==='alt'){
                        var seq='\\x1b'+e.key;if(window._wsSend)window._wsSend(seq);else if(window.term)window.term.input(seq);
                    }
                }
            },true);
        })();
        """

        func webView(_ webView: WKWebView, didFinish navigation: WKNavigation!) {
            print("[TerminalWebView] Loaded: \(webView.url?.absoluteString ?? "nil")")
            // Delay to let xterm.js initialize
            DispatchQueue.main.asyncAfter(deadline: .now() + 1.0) {
                webView.evaluateJavaScript(Self.setupJS, completionHandler: nil)
            }
        }

        func webView(_ webView: WKWebView, didFailProvisionalNavigation navigation: WKNavigation!, withError error: Error) {
            print("[TerminalWebView] Failed: \(error.localizedDescription)")
        }
    }
}

File: android/ClaudeCursor/app/src/main/java/com/claudecursor/app/ui/components/TerminalWebView.kt
package com.claudecursor.app.ui.components

import android.annotation.SuppressLint
import android.webkit.WebSettings
import android.webkit.WebView
import android.webkit.WebViewClient
import androidx.compose.runtime.*
import androidx.compose.ui.Modifier
import androidx.compose.ui.viewinterop.AndroidView
import com.claudecursor.app.ui.screens.sessionlist.SessionListViewModel

private const val USER_AGENT =
    "Mozilla/5.0 (Linux; Android 14; Pixel 8) AppleWebKit/537.36 (KHTML, like Gecko) " +
    "Chrome/120.0.0.0 Mobile Safari/537.36 ClaudeCursorCompanion/1.0"

/** JS injected after page load to fix heights and disable broken WebGL */
private const val FIX_PAGE_JS = """
(function(){
    // Disable WebGL (broken on Android emulators)
    if(!window._webglDisabled){
        window._webglDisabled=true;
        var orig=HTMLCanvasElement.prototype.getContext;
        HTMLCanvasElement.prototype.getContext=function(type,attrs){
            if(type==='webgl'||type==='webgl2'||type==='experimental-webgl') return null;
            return orig.call(this,type,attrs);
        };
    }

    // Force pixel dimensions on xterm container (CSS 100% chain broken on Android WebView)
    var h=window.innerHeight, w=window.innerWidth;
    document.documentElement.style.cssText='height:'+h+'px;width:'+w+'px;margin:0;padding:0;overflow:hidden';
    document.body.style.cssText='height:'+h+'px;width:'+w+'px;margin:0;padding:0;overflow:hidden';
    var x=document.querySelector('.xterm');
    if(x) x.style.cssText='height:'+h+'px;width:'+w+'px;position:absolute;top:0;left:0';
    var els=document.querySelectorAll('#terminal-container,#terminal');
    for(var i=0;i<els.length;i++) els[i].style.cssText='height:'+h+'px;width:'+w+'px';
    console.log('[app] forced dimensions: '+w+'x'+h+', xterm='+(x?'found':'missing'));
})();
"""

/** JS injected after page load to fix fonts, set up modifier keys, and refresh terminal */
private const val SETUP_JS = """
(function(){
    // Font fix + resize: retry until window.term exists
    if(window.term){
        if(!window._fontFixed){
            window._fontFixed=true;
            try{
                window.term.options.fontFamily='monospace';
                console.log('[app] font fixed, rows='+window.term.rows+' cols='+window.term.cols);
            }catch(e){}
            window.dispatchEvent(new Event('resize'));
            setTimeout(function(){ window.dispatchEvent(new Event('resize')); },300);
        }
    } else {
        console.log('[app] term not ready, will retry');
    }

    // Modifier key listener: add only once
    if(!window._modKeySetup){
        window._modKeySetup=true;
        window._pendingModifier=null;
        document.addEventListener('keydown',function(e){
            if(!window._pendingModifier||e.key.length!==1) return;
            e.preventDefault();
            e.stopImmediatePropagation();
            var mod=window._pendingModifier;
            window._pendingModifier=null;
            if(window.term){
                if(mod==='ctrl'){
                    var c=e.key.toUpperCase().charCodeAt(0)-64;
                    if(c>0&&c<27){var ch=String.fromCharCode(c);if(window._wsSend)window._wsSend(ch);else if(window.term)window.term.input(ch);}
                } else if(mod==='alt'){
                    var seq='\x1b'+e.key;if(window._wsSend)window._wsSend(seq);else if(window.term)window.term.input(seq);
                }
            }
        },true);
    }
})();
"""

@SuppressLint("SetJavaScriptEnabled")
@Composable
fun TerminalWebView(
    baseURL: String,
    sessionName: String,
    sharePath: String? = null,
    viewModel: SessionListViewModel,
    modifier: Modifier = Modifier
) {
    val path = sharePath ?: "/terminal/$sessionName/"
    val url = "${baseURL.trimEnd('/')}$path"

    var currentSession by remember { mutableStateOf("") }

    AndroidView(
        factory = { context ->
            WebView.setWebContentsDebuggingEnabled(true)
            WebView(context).apply {
                settings.apply {
                    javaScriptEnabled = true
                    domStorageEnabled = true
                    userAgentString = USER_AGENT
                    mediaPlaybackRequiresUserGesture = false
                    @SuppressLint("SetJavaScriptEnabled")
                    cacheMode = WebSettings.LOAD_DEFAULT
                    useWideViewPort = false
                    loadWithOverviewMode = false
                }

                webViewClient = object : WebViewClient() {
                    override fun onPageFinished(view: WebView?, pageUrl: String?) {
                        super.onPageFinished(view, pageUrl)
                        evaluateJavascript(FIX_PAGE_JS, null)
                        evaluateJavascript(SETUP_JS, null)
                        for (delay in longArrayOf(500, 1500, 3000)) {
                            postDelayed({
                                evaluateJavascript(FIX_PAGE_JS, null)
                                evaluateJavascript(SETUP_JS, null)
                            }, delay)
                        }
                    }
                }

                setBackgroundColor(android.graphics.Color.BLACK)
                isVerticalScrollBarEnabled = false
                isHorizontalScrollBarEnabled = false

                viewModel.webView = this
                loadUrl(url)
                currentSession = sessionName
            }
        },
        update = { webView ->
            if (currentSession != sessionName) {
                val newUrl = "${baseURL.trimEnd('/')}${sharePath ?: "/terminal/$sessionName/"}"
                webView.loadUrl(newUrl)
                currentSession = sessionName
                viewModel.webView = webView
            }
        },
        modifier = modifier
    )
}
